tests/core/contracts/test_contract_util_functions.py

```diff
@@ -1,3 +1,9 @@
+from web3.contract import (
+    parse_block_identifier_int,
+)
+
+
 def test_parse_block_identifier_int(web3):
     last_num = web3.eth.getBlock('latest').number
-    assert web3.eth.getBlock(0) == web3.eth.getBlock(-1 - last_num)
+    assert web3.eth.getBlock(0) == web3.eth.getBlock(
+        parse_block_identifier_int(web3, -1 - last_num))
```

Comment:
web3.eth.getBlock doesnt accept negative block identifiers.  parse_block_identifier_int will translate negative identifiers into the block number.  Maybe the assertion should just be:

```
assert 0 == parse_block_identifier_int(web3, -1 - last_num)
```

Reply:
It's very strange that `parse_block_identifier` would normalize a negative number to zero.  That strikes me as unexpected behavior and likely to result in accidental errors.  It seems like if the number if negative then it should raise a `ValidationError` or something like that.

Reply:
`parse_block_identifier` has logic to handle negative block numbers in a similar way Python does it for lists, e.g. `parse_block_identifier(-2)` will return the block before the last one. 
Currently this method would call getBlock('latest') for every invocation of a contract. What this pull request does is eliminates this extra call when block identifier (for the contract function call) is not numeric, e.g. 'latest'.
In my opinion, handling negative block identifiers for contract function calls should be removed for 2 reasons: (1) as @pipermerriam said it's unexpected behavior, (2) it's inconsistent with `web3.getBlock()` which doesn't handle negative block numbers. The problem with removing this logic is it will break backward compatibility.

Reply:
Yeah, it seems to have been added in https://github.com/ethereum/web3.py/pull/663/commits/9f23db085e2405dd29c280f1b9ca5c6c65ee93f8

I don't see any rationale in the related PR or issue for why a negative index was particularly helpful or important. I'm happy to see it get dropped in v5. I also understand why this test did what it did now. So :+1: from me to merge this PR (which isn't making the situation worse) as is. 

Although it's probably worth adding a comment on the test, mentioning that it is converting the negative index from the latest block back to zero. Also saying that the test has to do the conversion explicitly because transaction calls allow this negative index, but `getBlock()` doesn't.